cosmovisor/upgrade_test.go

```diff
@@ -0,0 +1,285 @@
+	require.NoError(t, err)
+	defer os.RemoveAll(home)
+
+	cfg := Config{Home: home, Name: "dummyd"}
+
+	currentBin, err := cfg.CurrentBin()
+	require.NoError(t, err)
+	assert.Equal(t, cfg.GenesisBin(), currentBin)
+	assertCurrentLink(t, cfg, "genesis")
+
+	err = cfg.SetCurrentUpgrade("chain2")
+	require.NoError(t, err)
+	currentBin, err = cfg.CurrentBin()
+	require.NoError(t, err)
+	assert.Equal(t, cfg.UpgradeBin("chain2"), currentBin)
+	assertCurrentLink(t, cfg, filepath.Join("upgrades", "chain2"))
+}
+
+func assertCurrentLink(t *testing.T, cfg Config, target string) {
+	link := filepath.Join(cfg.Root(), currentLink)
+	// ensure this is a symlink
+	info, err := os.Lstat(link)
+	require.NoError(t, err)
+	require.Equal(t, os.ModeSymlink, info.Mode()&os.ModeSymlink)
+
+	dest, err := os.Readlink(link)
+	require.NoError(t, err)
+	expected := filepath.Join(cfg.Root(), target)
+	require.Equal(t, expected, dest)
+}
+
+// TODO: test with download (and test all download functions)
+func TestDoUpgradeNoDownloadUrl(t *testing.T) {
+	home, err := copyTestData("validate")
+	require.NoError(t, err)
+	defer os.RemoveAll(home)
+
+	cfg := &Config{Home: home, Name: "dummyd", AllowDownloadBinaries: true}
+
+	currentBin, err := cfg.CurrentBin()
+	require.NoError(t, err)
+
+	assert.Equal(t, cfg.GenesisBin(), currentBin)
+
+	// do upgrade ignores bad files
+	for _, name := range []string{"missing", "nobin", "noexec"} {
+		info := &UpgradeInfo{Name: name}
+		err = DoUpgrade(cfg, info)
+		require.Error(t, err, name)
+		currentBin, err := cfg.CurrentBin()
+		require.NoError(t, err)
+		assert.Equal(t, cfg.GenesisBin(), currentBin, name)
+	}
+
+	// make sure it updates a few times
+	for _, upgrade := range []string{"chain2", "chain3"} {
+		// now set it to a valid upgrade and make sure CurrentBin is now set properly
+		info := &UpgradeInfo{Name: upgrade}
+		err = DoUpgrade(cfg, info)
+		require.NoError(t, err)
+		// we should see current point to the new upgrade dir
+		upgradeBin := cfg.UpgradeBin(upgrade)
+		currentBin, err := cfg.CurrentBin()
+		require.NoError(t, err)
+
+		assert.Equal(t, upgradeBin, currentBin)
+	}
+}
+
+func TestOsArch(t *testing.T) {
+	// all download tests will fail if we are not on linux...
+	assert.Equal(t, "linux/amd64", osArch())
+}
+
+func TestGetDownloadURL(t *testing.T) {
+	// all download tests will fail if we are not on linux...
+	ref, err := filepath.Abs(filepath.FromSlash("./testdata/repo/ref_zipped"))
+	require.NoError(t, err)
+	badref, err := filepath.Abs(filepath.FromSlash("./testdata/repo/zip_binary/autod.zip"))
+	require.NoError(t, err)
+
+	cases := map[string]struct {
+		info  string
+		url   string
+		isErr bool
+	}{
+		"missing": {
+			isErr: true,
+		},
+		"follow reference": {
+			info: ref,
+			url:  "https://github.com/cosmos/cosmos-sdk/blob/joanthan/integrate-cosmosd/cosmovisor/testdata/repo/zip_directory/autod.zip?checksum=sha256:3784e4574cad69b67e34d4ea4425eff140063a3870270a301d6bb24a098a27ae",
```

Comment:
Nice, this one is using refs to this repo.

You made a choice to tag to commits not branches? Not sure if this is better or worse (for causing breakage)

Reply:
Branch tests will definitely fail when merged to master and the branch deleted (you will need a quick follow up PR pinning the tests to master branch not joanthan)

Reply:
Yeah I was thinking to do that, since in the end master is a branch that for sure it will exist. So I will create a fast PR once merged. :D